packages/instant/src/util/heap.ts

```diff
@@ -0,0 +1,115 @@
+
+import { ANALYTICS_ENABLED, HEAP_ANALYTICS_DEVELOPMENT_APP_ID, HEAP_ANALYTICS_PRODUCTION_APP_ID } from '../constants';
+
+import { AnalyticsEventOptions, AnalyticsUserOptions } from './analytics';
+
+export interface HeapAnalytics {
+    loaded: boolean;
+    appid: string;
+    identify(id: string, idType: string): void;
+    track(eventName: string, eventProperties?: ObjectMap<string | number>): void;
+    resetIdentity(): void;
+    addUserProperties(properties: AnalyticsUserOptions): void;
+    addEventProperties(properties: AnalyticsEventOptions): void;
+    removeEventProperty(property: string): void;
+    clearEventProperties(): void;
+}
+interface ModifiedWindow {
+    heap?: HeapAnalytics;
+    zeroExInstantLoadedHeap?: boolean;
+}
+const getWindow = (): ModifiedWindow => {
+    return window as ModifiedWindow;
+};
+
+const getHeapAppId = (): string => {
+    if (process.env.NODE_ENV === 'production') {
+        return HEAP_ANALYTICS_PRODUCTION_APP_ID;
+    }
+    return HEAP_ANALYTICS_DEVELOPMENT_APP_ID;
+};
+
+const setupZeroExInstantHeap = () => {
+    const curWindow = getWindow();
+    // Set property to specify that this is zeroEx's heap
+    curWindow.zeroExInstantLoadedHeap = true;
+
+    // Typescript-compatible version of https://docs.heapanalytics.com/docs/installation
+    /* tslint:disable */
+    ((window as any).heap = (window as any).heap || []),
+        ((window as any).heap.load = function(e: any, t: any) {
+            ((window as any).heap.appid = e), ((window as any).heap.config = t = t || {});
+            var r = t.forceSSL || 'https:' === (document.location as Location).protocol,
+                a = document.createElement('script');
+            (a.type = 'text/javascript'),
+                (a.async = !0),
+                (a.src = (r ? 'https:' : 'http:') + '//cdn.heapanalytics.com/js/heap-' + e + '.js');
+            var n = document.getElementsByTagName('script')[0];
+            (n.parentNode as Node).insertBefore(a, n);
+            for (
+                var o = function(e: any) {
+                        return function() {
+                            (window as any).heap.push([e].concat(Array.prototype.slice.call(arguments, 0)));
+                        };
+                    },
+                    p = [
+                        'addEventProperties',
+                        'addUserProperties',
+                        'clearEventProperties',
+                        'identify',
+                        'resetIdentity',
+                        'removeEventProperty',
+                        'setEventProperties',
+                        'track',
+                        'unsetEventProperty',
+                    ],
+                    c = 0;
+                c < p.length;
+                c++
+            )
+                (window as any).heap[p[c]] = o(p[c]);
+        });
+    (window as any).heap.load(getHeapAppId());
+    /* tslint:enable */
+
+    return curWindow.heap as HeapAnalytics;
```

Comment:
There is a bit of a race condition here. What I remember happening is that `heap` is initialized to this thing with the `load` method on it, and then you call load, and then that method will replace `heap` with something that does adhere to `HeapAnalytics` at some point in the future. However, in the mean-time, it does not adhere to the interface and only has the `loaded` boolean I believe(?).

Reply:
I think this is safe.  Here's what I observed from experimenting with this.

My code was as follows:

```js
<...heap integration snippet>
console.log("beforeLoad", heap.addEventProperties);
heap.load("YOUR_APP_ID");
console.log("afterLoad", heap.addEventProperties);
```

The last `console.log` was executed immediately after the `.load` call was made, and before heap actually loaded their full JS.  The output was as follows:

![image](https://user-images.githubusercontent.com/210914/48653422-eed5a300-e9b9-11e8-88c0-b183e3a9b2d0.png)

So it looks like it's pretty smart, which is what I would expect from an analytics snippet like this: immediately after `.load` it makes the public functions available: they push the data into an array which is later consumed by the full script once it loads from the CDN.

Let me know if I'm missing something

Source: https://gist.github.com/steveklebanoff/3de834cf8f8d2e30cceace4a87cf3aed